oauth2_http/java/com/google/auth/oauth2/ImpersonatedCredentials.java

```diff
@@ -0,0 +1,405 @@
+   *             will either be deleted or made private in a later version.
+   */
+  @Deprecated
+  ImpersonatedCredentials(GoogleCredentials sourceCredentials, String targetPrincipal, List<String> delegates,
+      List<String> scopes, int lifetime, HttpTransportFactory transportFactory) {
+    this.sourceCredentials = sourceCredentials;
+    this.targetPrincipal = targetPrincipal;
+    this.delegates = delegates;
+    this.scopes = scopes;
+    this.lifetime = lifetime;
+    this.transportFactory = firstNonNull(transportFactory,
+        getFromServiceLoader(HttpTransportFactory.class, OAuth2Utils.HTTP_TRANSPORT_FACTORY));
+    this.transportFactoryClassName = this.transportFactory.getClass().getName();
+    if (this.delegates == null) {
+      this.delegates = new ArrayList<String>();
+    }
+    if (this.scopes == null) {
+      throw new IllegalStateException(SCOPE_EMPTY_ERROR);
+    }
+    if (this.lifetime > ONE_HOUR_IN_SECONDS) {
+      throw new IllegalStateException(LIFETIME_EXCEEDED_ERROR);
+    }
+  }
+
+  /**
+   * @param sourceCredentials = Source Credentials.
+   * @param targetPrincipal   = targetPrincipal;
+   * @param scopes            = scopes;
+   * @param lifetime          = lifetime;
+   * @param transportFactory  = HTTP transport factory, creates the transport used
+   *                          to get access tokens.
+   * @deprecated Use {@link #create(ImpersonatedCredentials)} instead. This constructor
+   *             will either be deleted or made private in a later version.
+   */  
+  @Deprecated
+  ImpersonatedCredentials(GoogleCredentials sourceCredentials, String targetPrincipal, List<String> scopes,
+      int lifetime, HttpTransportFactory transportFactory) {
+    this(sourceCredentials, targetPrincipal, new ArrayList<String>(), scopes, lifetime, transportFactory);
+  }
+
+  /**
+   * @param sourceCredentials = Source Credentials.
+   * @param targetPrincipal   = targetPrincipal;
+   * @param delegates         = delegates;
+   * @param scopes            = scopes;
+   * @param lifetime          = lifetime;
+   * @deprecated Use {@link #create(ImpersonatedCredentials)} instead. This constructor
+   *             will either be deleted or made private in a later version.
+   */  
+  @Deprecated
+  ImpersonatedCredentials(GoogleCredentials sourceCredentials, String targetPrincipal, List<String> scopes,
+      int lifetime) {
+    this(sourceCredentials, targetPrincipal, new ArrayList<String>(), scopes, lifetime, null);
+  }
+
+  @Override
+  public AccessToken refreshAccessToken() throws IOException {
+    if (this.sourceCredentials.getAccessToken() == null) {
+      this.sourceCredentials = this.sourceCredentials.createScoped(Arrays.asList(CLOUD_PLATFORM_SCOPE));
+      this.sourceCredentials.refresh();
+    }
+    if (this.sourceCredentials.getAccessToken().getExpirationTime().before(new Date())) {
+      this.sourceCredentials.refresh();
```

Comment:
It might also be nice to wrap an IOException here to qualify that it comes from refreshing the source credentials. This can also be added at a later time.

Reply:
as above, bubbled `IOException`